fe/fe-core/src/main/java/com/starrocks/catalog/StarOSAgent.java

```diff
@@ -226,13 +255,22 @@ private List<ReplicaInfo> getShardReplicas(long shardId) throws UserException {
     public long getPrimaryBackendIdByShard(long shardId) throws UserException {
         List<ReplicaInfo> replicas = getShardReplicas(shardId);
 
-        try (LockCloseable lock = new LockCloseable(rwLock.readLock())) {
+        try (LockCloseable lock = new LockCloseable(rwLock.writeLock())) {
             for (ReplicaInfo replicaInfo : replicas) {
                 if (replicaInfo.getReplicaRole() == ReplicaRole.PRIMARY) {
                     WorkerInfo workerInfo = replicaInfo.getWorkerInfo();
                     long workerId = workerInfo.getWorkerId();
                     if (!workerToBackend.containsKey(workerId)) {
-                        throw new UserException("Failed to get backend by worker. worker id: " + workerId);
+                        // get backendId from system info
+                        String workerAddr = workerInfo.getIpPort();
+                        String[] pair = workerAddr.split(":");
+                        long backendId = GlobalStateMgr.getCurrentSystemInfo().getBackendIdByHost(pair[0]);
```

Comment:
It is possible to deploy multiple BEs on a machine.
Should use `host` and `starlet_port`, and check whether `backendId` is -1.